[AutoDiff] Forward-mode support for variables, generics, tuples, structs.

bartchr808 opened the pull request:
This PR will support everything other than control flow (coming in a PR later next week 🚀) and arrays for forward mode differentiation that was initially supported in https://github.com/apple/swift/pull/26057.

Commit history is completely messed up due to having to work off separate branches due to some progress blocking bug problems unfortunately 😄

bartchr808 committed:
Merge in work from branch jvp-emitter.

bartchr808 committed:
Init commit.

bartchr808 committed:
WIP.

bartchr808 committed:
Make DifferentialEmitter create differential.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Typos and small bugs.

bartchr808 committed:
Get visitApplyInst and visitReturnInst working.

bartchr808 committed:
Remove multi block logic.

bartchr808 committed:
WIP

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Accidentally removed mapper.

bartchr808 committed:
Move JVP lower and start adding AdjointValue.

bartchr808 committed:
Create differential builder.

bartchr808 committed:
Get value mapping for seed params.

bartchr808 committed:
PullbackInfo -> LinearMapInfo.

bartchr808 committed:
Get correct output, have retain memory leak.

bartchr808 committed:
Move around code.

bartchr808 committed:
Fix release bug.

bartchr808 committed:
More cleanup.

bartchr808 committed:
More cleanup x2.

bartchr808 committed:
Add tangent accumulation.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
PR feedback #1.

bartchr808 committed:
WIP: start adding tests.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Fix SIL tests.

bartchr808 committed:
Remove tangent aggregation logic and cleanup.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Fix comments, add jvpNegate, add/modify tests, simplify lit flag.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Merge in work from var and struct support.

bartchr808 committed:
Get diff struct bb mapping fixed.

bartchr808 committed:
WIP: handle tuple inst.

bartchr808 committed:
PR feedback.

bartchr808 committed:
Update new test due to upstream changes and existing tests now that JVP is generated.

bartchr808 committed:
PR feedback.

bartchr808 committed:
Fix tests and PR feedback.

bartchr808 committed:
PR feedback.

bartchr808 committed:
Merge in changes from intial DifferentialEmitter PR.

dan-zheng committed:
[NFC] Fix naming.

"linear map enum" -> "branching trace enum"
Other assorted gardening.

bartchr808 committed:
PR feedback and additional 'Tracked<Float>' tests.

bartchr808 committed:
[NFC] Small name and spacing changes.

bartchr808 committed:
Merge in changes from base branch 'differential-emitter'.

bartchr808 committed:
Get generics almost working.

bartchr808 committed:
Create the differential struct with func reabstraction.

bartchr808 committed:
Start adding tests.

bartchr808 committed:
WIP: more tests.

bartchr808 committed:
Get TF-508 and TF-523 tests case working.

bartchr808 committed:
Get TF-534 working.

bartchr808 committed:
Spacing.

bartchr808 committed:
WIP: handle tuples with Int element.

bartchr808 committed:
Add tuple extract support, TODO tests.

bartchr808 committed:
WIP: array init.

bartchr808 committed:
Get array init working with Xcode.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Revamp linear map info struct creation.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter

bartchr808 committed:
Merge branch 'differential-emitter' into differential-emitter-generic-support

bartchr808 committed:
WIP: fix generics

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter-generic-support

bartchr808 committed:
Squash merge in latest changes from differential-emitter.

bartchr808 committed:
Squash merge in latest changes from differential-emitter.

bartchr808 committed:
WIP

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter-generic-support

bartchr808 committed:
Make updates for ownership and handle subset autodiff params.

bartchr808 committed:
Cleanup.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4964,13 +5392,205 @@ class JVPEmitter final
+        ri->getLoc(), joinElements(directResults, builder, loc));
+  }
+
+  void visitBeginBorrowInst(BeginBorrowInst *bbi) {
```
These things are super boilerplate-y. Could you define a macro (like `NOT_DIFFERENTIABLE`) instead?

```cpp
#define CLONE_AND_EMIT_TANGENT(INST, ID) \
  void visit##INST##Inst(INST##Inst *inst) { \
    TypeSubstCloner::visit##INST##Inst(inst); \
    if (shouldBeDifferentiated(inst, getIndices())) \
      emitTangentFor##Inst(inst); \
  } \
  void emitTangentFor##INST(INST##Inst *(ID))

CLONE_AND_EMIT_TANGENT(BeginBorrow, bbi) {
  // Add your begin_borrow handling code.
}

CLONE_AND_EMIT_TANGENT(EndBorrow, ebi) {
  // Add your end_borrow handling code.
}

#endif // CLONE_AND_EMIT_TANGENT
```

Note: I didn't spend time to make sure my code above is syntactically correct.

bartchr808 replied:
Ya definitely agree! Will add it once I finish porting over the existing protocol tests from `protocol_requiremenet_autodiff.swift`

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4388,27 +4444,51 @@ class JVPEmitter final
+        auto val = materializeTangent(diffVal, diffLoc);
+        retElts.push_back(val);
+    };
+    // Collect differentiation parameter adjoints.
```
Not sure I understand this comment. Doesn't seem to be what this code is doing.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4449,8 +4558,450 @@ class JVPEmitter final
+    // Get the Tangent of the operand (the struct buffer).
+    auto tanOperand = getTangentBuffer(bb, seai->getOperand());
+
+    // Emit the instruction
```
```suggestion
    // Emit the instruction.
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4417,7 +4497,18 @@ class JVPEmitter final
-        tanParam = materializeTangent(getTangentValue(origArg), loc);
+        // If original result is a `tuple_extract`, materialize tangent value of
+        // `ai` and extract the corresponding element tangent value.
+        if (auto *tupleExtract =
```
Same comment as https://github.com/apple/swift/pull/26057#discussion_r315466008.

bartchr808 committed:
Fix tuple support.

bartchr808 committed:
merge partial revamp.

bartchr808 committed:
Merge branch 'tensorflow' into differential-emitter-generic-support

bartchr808 committed:
WIP: activity analysis revamp.

bartchr808 committed:
Update broken tests with latest activity analysis and ownership changes.

bartchr808 commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1343,6 +1377,10 @@ using Activity = OptionSet<ActivityFlags>;
 /// indices.
 class DifferentiableActivityInfo {
 private:
+  // TODO: temporary hack while we investigate why reverse mode doesn't work
```
Just so it's easier to go back and remove the temporary work around, wherever this comment exists, it should be removed once we figure out how we can use `shouldDifferentiateApplyInst` without special casing for forward or reverse mode. Let me know if there is a better way to mark these locations!

dan-zheng replied:
Filed [TF-800](https://bugs.swift.org/browse/TF-800) to track activity analysis and "should differentiate" logic differences between forward- and reverse-mode. Edited todo comments to reference TF-800 in https://github.com/apple/swift/pull/26743/commits/9ed94482aa1e62884d3e73f8e3d878e379079f9d.

bartchr808 committed:
Gardening.

bartchr808 commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3606,7 +3741,21 @@ class VJPEmitter final
+        allResults.push_back(ai);
+      }
+    } else {
+      allResults.push_back(ai);
```
This has the duplicate line inside the if above when it's a tuple type, let me know if I should reformat this to avoid it, or if this is fine.

dan-zheng replied:
Reformatted in https://github.com/apple/swift/pull/26743/commits/c5ebacc9df98d60f6f52d1c834d11c2a1b5a3620 by using helper boolean flags, e.g. [`bool isDifferentialAndFoundDestructureTupleUser`](https://github.com/apple/swift/pull/26743/commits/c5ebacc9df98d60f6f52d1c834d11c2a1b5a3620#diff-4ede198777b59748c377cdd17c444e83R1656).

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -122,6 +142,24 @@ collectAllFormalResultsInTypeOrder(SILFunction &function,
+  auto *retInst = cast<ReturnInst>(function.findReturnBB()->getTerminator());
+  auto retVal = retInst->getOperand();
+  if (auto *tupleInst =
+      dyn_cast_or_null<TupleInst>(retVal->getDefiningInstruction()))
```
Simplify and move to the end of the previous line.
```suggestion
      dyn_cast<TupleInst>(retVal))
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4378,44 +4572,134 @@ class JVPEmitter final
-    auto loc = ri->getOperand().getLoc();
-    auto diffBuilder = getDifferentialBuilder();
+  /// Helper that emits the final return instruction in a basic block of the
+  /// differential function. This is needed for when there are indrect outputs.
```
> This is needed for when there are indrect outputs.

Isn't this needed for all cases?

bartchr808 replied:
Oh yup old comment forgot to update!

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+        for (auto use : ai->getUses()) {
+          if (auto *dti = dyn_cast<DestructureTupleInst>(use->getUser())) {
+            numberOfDestructureTupleInsts++;
+            assert(numberOfDestructureTupleInsts == 1 && "There should only "
```
```suggestion
            assert(numDestructureTupleInsts == 1 && "There should only "
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+          if (auto *dti = dyn_cast<DestructureTupleInst>(use->getUser())) {
+            numberOfDestructureTupleInsts++;
+            assert(numberOfDestructureTupleInsts == 1 && "There should only "
+                   "be one use of the tuple in a 'destructure_tuple' inst.");
```
Rewording a bit to make clear
```suggestion
                   "be one use of the tuple that is a 'destructure_tuple' inst.");
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+
+  /// Handle `struct_extract` instruction.
+  ///   Original: y = struct_extract x, #field
+  ///   Tangent: tan[y] = struct_extract tan[x], tan[#field]]
```
```suggestion
  ///   Tangent: tan[y] = struct_extract tan[x], tan[#field]
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+
+  /// Handle `struct_element_addr` instruction.
+  ///   Original: y = struct_element_addr x, #field
+  ///   Tangent: tan[y] = struct_element_addr tan[x], tan[#field]]
```
```suggestion
  ///   Tangent: tan[y] = struct_element_addr tan[x], tan[#field]
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+    }
+  }
+
+#undef CLONE_AND_EMIT_TANGENT
```
```suggestion
#undef CLONE_AND_EMIT_TANGENT

```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+    auto tanTuple =
+        materializeTangent(getTangentValue(dti->getOperand()), loc);
+    auto *tupleElements = diffBuilder.createDestructureTuple(loc, tanTuple);
+    for (auto i : range(tupleElements->getResults().size())) {
```
```suggestion
    for (auto i : range(tupleElements->getNumResults())) {
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+    auto *bb = dti->getParent();
+    auto loc = dti->getLoc();
+
+    SmallVector<SILValue, 2> activeOrigResult;
```
```suggestion
    SmallVector<SILValue, 2> activeOrigResults;
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+        break;
+      }
+    }
+    assert(activeOrigResult.size() > 0 &&
```
```suggestion
    assert(!activeOrigResult.empty() &&
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+
+    SmallVector<SILValue, 2> activeOrigResult;
+    bool hasActiveResult = false;
+    for (auto i : range(dti->getResults().size())) {
```
```suggestion
    for (auto result : dti->getResults()) {
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+    assert(origIndResults.size() == diffIndResults.size());
+
+    for (auto &origBB : *original) {
+      for (auto i : range(diffIndResults.size())) {
```
```suggestion
      for (auto i : indices(diffIndResults)) {
```

or if not compiling,

```suggestion
      for (auto i : swift::indices(diffIndResults)) {
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
+        unsigned numberOfDestructureTupleInsts = 0;
+        bool notSetValue = true;
+        for (auto use : ai->getUses()) {
+          if (auto *dti = dyn_cast<DestructureTupleInst>(use->getUser())) {
```
Shouldn't you use the `getSingleDestructureTupleInstUser` helper?

bartchr808 replied:
Ahh yup accidentally put some of the `There should only be one use of the tuple that is a 'destructure_tuple' inst` verification here as well, moved it to `getSingleDestructureTupleUser` and called that function here instead. Still kept the `should only have one active result from the tuple logic` here though.

bartchr808 committed:
PR feedback.

bartchr808 commented:
@swift-ci please test tensorflow

bartchr808 commented:
@swift-ci please test tensorflow

bartchr808 commented:
@rxwei @dan-zheng just so it's easier for you to merge it in once we have time, spent a minute to deal with the merge conflict and verify it still works 😄

bartchr808 commented:
@swift-ci please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1546,17 +1610,59 @@ bool LinearMapInfo::shouldDifferentiateInstruction(SILInstruction *inst) {
+  // with us handling `apply` instructions differently.
+  switch (kind) {
+  case AutoDiffLinearMapKind::Differential: {
+    CHECK_INST_TYPE_ACTIVE_DEST(StoreInst)
```
It's better to define this kind of macros immediately before you use them and undefine immediately after.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/26743/commits/73f117604934b50ac600be3c9713fa39436d9d8a. Macro `#define`/`#undef` are tight around the use site.

dan-zheng committed:
Disabled non-varied result warning.

Follow-up to https://github.com/apple/swift/pull/26928 for reverse-mode.
TF-788 tracks re-enabling the warning.

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into differential-emitter-generic-support

dan-zheng committed:
Differential generation fixes.

- Handle `copy_value` and `destroy_addr` in "should differentiate" logic.
  These should be visited if they have an active operand.
- Add differential visitors for `destroy_value` and `destroy_addr`.

dan-zheng committed:
Cleanup: simplify `getSingleDestructureTupleUser`.

Remove precondition that argument must be tuple-typed to simplify call sites.
Clean up call site code and doc comments.

dan-zheng committed:
Cleanup: add todo comments referencing TF-800.

TF-800 tracks special-case activity analysis and "should differentiate" logic
regarding tuple-typed `apply` results.

dan-zheng committed:
Assorted cleanup.

dan-zheng commented:
I added:
* Fixes after merging from `tensorflow`: https://github.com/apple/swift/pull/26743/commits/82e9605dc8993fc023efee6c852c47935ec9c37a, https://github.com/apple/swift/pull/26743/commits/73f117604934b50ac600be3c9713fa39436d9d8a.
* Various cleanup.

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
Merging to unblock high priority patch https://github.com/apple/swift/pull/26709.
Happy to address feedback in a follow-up. There's more room for cleanup.
